### ISSUE TITLE
Vim: Reading from stdin...

### ISSUE THREAD

**Issue by ghost** [vim/vim#1552]
~~~c
#ifndef ALWAYS_USE_GUI
  mch_msg(_("Vim: Reading from stdin...\n"));
#endif
~~~

http://github.com/vim/vim/blob/47ffb90/src/fileio.c#L830-L832

Shouldn’t this be printing to stderr? Otherwise it prevents command like this
from being of any use:

~~~sh
$ echo hello | vi - -esc '%p|q!'
Vim: Reading from stdin...
hello
~~~

Currently you have to use /dev/stdin to silence the message:

~~~sh
$ echo hello | vi -esc '%p|q' /dev/stdin
hello
~~~

Or hi-jack /dev/stderr:

~~~sh
$ echo hello | vi - -esc 'wq!/dev/stderr' 2>&1 >/dev/null
hello
~~~

Ideally the message could be totally suppressed with -V0 or similar, but stderr
would be a start.

http://vi.stackexchange.com/q/4682/suppress-the-reading-from-stdin-message

**Comment by brammool** [comment]
Steven Penny wrote:

> ~~~c
> #ifndef ALWAYS_USE_GUI
>   mch_msg(_("Vim: Reading from stdin...\n"));
> #endif
> ~~~
> 
> http://github.com/vim/vim/blob/47ffb90/src/fileio.c#L830-L832
> 
> Shouldn’t this be printing to stderr? Otherwise it prevents command
> like this
> from being of any use:
> 
> ~~~sh
> $ echo hello | vi - -esc '%p|q!'
> Vim: Reading from stdin...
> hello
> ~~~
> 
> Currently you have to hi-jack stderr to silence the message:
> 
> ~~~sh
> $ echo hello | vi - -esc 'wq!/dev/stderr' 2>&1 >/dev/null
> hello
> ~~~

When starting a GUI program from the desktop or any other way that's not
in a terminal, the warning goes missing.  The idea is that Vim shows it
in the GUI, but that can happen only after it has started up.

Dropping the message causes other problems, such as the user not being
aware why Vim is stuck waiting.

-- 
Laughing helps. It's like jogging on the inside.

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by svnpenn** [comment]
> Dropping the message causes other problems, such as the user not being aware
> why Vim is stuck waiting.

@brammool I am not asking to drop the message. I am asking for one or more of
these:

- Put message though STDERR instead of STDOUT
- Allow suppression of message with -V0 or similar

**Comment by svnpenn** [comment]
https://github.com/vim/vim/pull/1564

**Comment by svnpenn** [comment]
https://github.com/neovim/neovim/pull/6298

**Comment by svnpenn** [comment]
@brammool notice carefully that NeoVim has removed this message:

https://github.com/neovim/neovim/commit/bdcb2a38b3667c01efada0c6ed9414426f1c0468

Will you consider this?

**Comment by msva** [comment]
He's just answered about it few comments above 😎 
And, AFAIK, it is Vim is upstream for NeoVim, but not vice versa.
So, 2+2 = I very doubt @brammool will consider dropping it just because NeoVim considered.

And, btw, I still having problems on getting rid of unneded newline after hiding this message in Vim, and also I still very busy at work, so have no time ATM to implement what @brammool asked in #1564 😒

**Comment by rdebath** [comment]
I'm not sure how dumb I'm being, but isn't the purpose of this message to tell some "luser" that they have to type something when they enter a bare `vim -` command.
That is:
 1) The message should only show if the user hasn't redirected stdin

``` 
$ gunzip -
gzip: compressed data not read from a terminal.
```
 2) The message should probably be more descriptive.
```
$ vim -
Vim: Reading from terminal, enter text then press Ctrl-D...

```

So this patch would do enough.
``` diff
diff --git a/src/fileio.c b/src/fileio.c
index 87a46b1..939403b 100644
--- a/src/fileio.c
+++ b/src/fileio.c
@@ -825,7 +825,7 @@ readfile(
         * still be running, don't move the cursor to the last line, unless
         * always using the GUI.
         */
-       if (read_stdin)
+       if (read_stdin && isatty(0))
        {
 #ifndef ALWAYS_USE_GUI
            mch_msg(_("Vim: Reading from stdin...\n"));
```

**Comment by tonymec** [comment]
I don't agree that the message should only appear if the user hasn't redirected stdin.

When I use Vim as a pager, by ending a command-line with `| view -`, the message `Reading from stdin…` confirms that I didn't goof, and that Vim is indeed going to display, from its stdin, the redirected output of the previous command.

Similarly, when I give two or more filenames on the Vim command-line, Vim tells me how many files it's going to edit. This is of course hidden when Console Vim swaps the main and alternate tty buffers, but it reappears at end-of-run. I like it.

Best regards,
Tony.

**Comment by rdebath** [comment]
What? That doesn't seem very useful. You never see the message until after (if you're using the alt-screen) because VIM wipes the screen ...
...
Ah, yes, VIM really isn't a good pager is it. That sort of slow trickle of status information from the pipeline should appear on the screen as it's running. Personally, I don't put that sort of stuff in a pager at all, that's what a 200000 line scrollback is for. But in the event that I'm unsure if I want to save it I may use `less` with it's follow and save commands. (and forget to redirect stderr too)

Anyway, assuming you're using `bash` as your line editor (most people seem to) a better way of slurping up a slow pipe would seem be to use something like `view <(sh doit)`.  But even with that you can't see it happening and you usually have the standard error redirection issue so a shell alias is needed to fix it too.

**Comment by eromoe** [comment]
I don't know why  vim can not edit file now. 

`deploy.sh` exists , but  type `vim deploy.sh`, just shows

![image](https://user-images.githubusercontent.com/3938751/32422085-126caa4c-c2d9-11e7-9412-1ce296bde2f9.png)

leave vim shows:
```
Vim: Reading from stdin...
```

**Comment by eromoe** [comment]
Never mind... comment out `plugins=(git)`  in ~/.zshrc  fix that ...

**Comment by kenorb** [comment]
@eromoe Your issue is different. Please create a new ticket for that. If it's solved, remove your messages.

**Comment by kenorb** [comment]
I also expect the message to come from stderr, so I expect the following command to suppress the warning:

```
$ echo hello | vi - -esc '%p|q!' 2> /dev/null
Vim: Reading from stdin...
hello
```

But the message is still there.

Another test:

```
$ echo hello | vi - -esc '%p|q!' > out.txt
$ cat out.txt 
Vim: Reading from stdin...
hello
```

Related:

- [Using `-` doesn't work as expected](https://github.com/vim/vim/issues/2342).
- neovim/neovim/pull/6298
- [How can I suppress the “reading from STDIN” message from within Vim?](https://vi.stackexchange.com/q/4682/467)
- [How to edit content from the standard input?](https://vi.stackexchange.com/q/737/467)
- [Vim: Reading from stdin](http://vim.1045645.n5.nabble.com/Vim-Reading-from-stdin-td1151013.html)

**Comment by kenorb** [comment]
As per this [comment](https://github.com/neovim/neovim/pull/6298#issuecomment-326792130):

> the purpose of this message to tell some "luser" that they have to type something when they enter a bare `vim -` command

So:

> The message should only show if the user hasn't redirected stdin

Ideally:

> The message should probably be more descriptive.

```
$ vim -
Vim: Reading from terminal, enter text then press Ctrl-D...
```

Suggested patch from @rdebath:

```
diff --git a/src/fileio.c b/src/fileio.c
index 87a46b1..939403b 100644
--- a/src/fileio.c
+++ b/src/fileio.c
@@ -825,7 +825,7 @@ readfile(
         * still be running, don't move the cursor to the last line, unless
         * always using the GUI.
         */
-       if (read_stdin)
+       if (read_stdin && isatty(0))
        {
 #ifndef ALWAYS_USE_GUI
            mch_msg(_("Vim: Reading from stdin...\n"));
```

**Comment by chrisbra** [comment]
It's a warning no error. So it is okay that the message occurs on stdout.

**Comment by svnpenn** [comment]
@chrisbra no, its not. if the warning comes through STDOUT, then users have no sane way of filtering the warning or the actual output

**Comment by chrisbra** [comment]
ah yes, however, note that vim is no batch processor and mainly intended for interactive use.

**Comment by svnpenn** [comment]
@chrisbra even if that is true (its not), the software should act in a way that will allow uses to filter the warning if they choose - currently it does not without ugly hacks described above

**Comment by chrisbra** [comment]
what is not true?

**Comment by svnpenn** [comment]
@chrisbra I am not going to get into a long back and forth with you on this. The simple fact is that currently warnings and regular output are both coming through STDOUT - until that is fixed (and it should be for obvious reasons) this is a good and valid issue and doesnt require any discussion from you on the validity of said issue

**Comment by chrisbra** [comment]
Look, vim is meant for interactive use. Using it as a stream editor is not its primary purpose. There are ways around it as you found out, but I do consider the warning to be useful in general case. Second, since Vim supports several platforms and GUIs, how does it affect GUIs? Will it then still be displayed to the user? I think it should.

**Comment by svnpenn** [comment]
> interactive use

you keep saying that, but its not germane to the discussion. as long a vim has a console use at all, then this is a valid issue - just because vim is mainly interactive (its not) doesnt mean that console issues shouldnt be fixed

> warning to be useful

yes the warning is useful, i never argued that (you seem to have a logic problem) - the issue is not if the warning is useful - the issue is "do people have a sane method to disable it, should they choose to? currently they dont

> how does it affect GUIs

who cares? this question isnt about the GUI - its about the console - and if my method of fixing this is implemented (console switch) - then GUI users will notices no difference

please its seems like you form multiple logical fallacies when making comments - perhaps you should consider that in the future

**Comment by chrisbra** [comment]
>  just because vim is mainly interactive (its not)

That is a bold statement. I think this is wrong.

> who cares? this question isnt about the GUI

We as developers care. Just because you don't doesn't make it invalid.

In general I think I don't like your bold tone. It is not going to convince. Please be nice. Thanks.

**Comment by svnpenn** [comment]
> That is a bold statement. I think this is wrong.

again, you are making a logical fallacy - it doesnt matter is VIM is 99% GUI use - as long as VIM has any console use at all - this is a good and valid issue and should be fixed

> We as developers care. Just because you don't doesn't make it invalid.

per my previous comment - the GUI will not be affected - perhaps you did not read all the way through my previous comment - i will post it again for your convenience:

if my method of fixing this is implemented (console switch) - then GUI users will notice no difference

> In general I think I don't like your bold tone. It is not going to convince. Please be nice. Thanks.

In general I dont like your tone of multiple logical fallacies - perhaps you could visit the below page for more information on how to stay - on topic in these types of discussion - thank you kindly

http://yourlogicalfallacyis.com

**Comment by kenorb** [comment]
The `ex` is symlinked to `vim` executable binary and sharing the same codebase. Please note that originally [`ex` editor](https://en.wikipedia.org/wiki/Ex_%28text_editor%29) was a command line oriented, then after given a full-screen visual interface become the `vi` text editor as per Wiki page:

> `ex` was eventually given a full-screen visual interface (adding to its command line oriented operation), thereby becoming the `vi` text editor.

So I think Vim (as **IM**proved version of **V**i) should support both interactive and non-interactive modes (Ex mode) equally as it's originally coming from non-interactive background as the history shows.

**Comment by kenorb** [comment]
Going back to the original problem, this command should work by using Ex editor (which suppose to be support non-interactive editing):

    echo hello | ex - -sc '%p|q!'

but it's not (prints nothing). Fixing this issue could solve the initial problem by using `ex` instead of `vim`. I've created for this separate issue at #2342

**Comment by chrisbra** [comment]
closed by https://github.com/vim/vim/commit/234d16286a2733adedef56784c17415ae169b9ad
